controllers/sensor/operator.go

```diff
@@ -90,126 +100,157 @@ func (soc *sOperationCtx) operate() error {
-			}
+		err := soc.updateSensorResources()
+		if err != nil {
+			return err
 		}
 
-		// Initialize all trigger nodes
-		for _, trigger := range soc.s.Spec.Triggers {
-			InitializeNode(soc.s, trigger.Name, v1alpha1.NodeTypeTrigger, &soc.log)
+	case v1alpha1.NodePhaseError:
+		soc.log.Info().Msg("sensor is in error state. check sensor resource status information and corresponding escalated K8 event for the error")
+
+		err := soc.updateSensorResources()
+		if err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		// add default env variables
-		soc.s.Spec.DeploySpec.Containers[0].Env = append(soc.s.Spec.DeploySpec.Containers[0].Env, []corev1.EnvVar{
-			{
-				Name:  common.SensorName,
-				Value: soc.s.Name,
-			},
-			{
-				Name:  common.SensorNamespace,
-				Value: soc.s.Namespace,
-			},
-			{
-				Name:  common.EnvVarSensorControllerInstanceID,
-				Value: soc.controller.Config.InstanceID,
-			},
-		}...,
-		)
+func (soc *sOperationCtx) createSensorResources() error {
+	err := ValidateSensor(soc.s)
+	if err != nil {
+		soc.log.Error().Err(err).Msg("failed to validate sensor")
+		soc.markSensorPhase(v1alpha1.NodePhaseError, false, err.Error())
+		return nil
+	}
 
-		// create a sensor pod
-		// default label on sensor
-		if soc.s.ObjectMeta.Labels == nil {
-			soc.s.ObjectMeta.Labels = make(map[string]string)
-		}
-		soc.s.ObjectMeta.Labels[common.LabelSensorName] = soc.s.Name
+	soc.initializeAllNodes()
+	pod, err := soc.createSensorPod()
+	if err != nil {
+		soc.log.Error().Err(err).Str("sensor-name", soc.s.Name).Msg("failed to create pod for sensor")
+		soc.markSensorPhase(v1alpha1.NodePhaseError, false, fmt.Sprintf("failed to create sensor pod. err: %s", err))
+		return err
+	}
+	soc.markAllNodePhases()
+	soc.log.Info().Str("pod-name", pod.Name).Msg("sensor pod is created")
 
-		sensorPod := &corev1.Pod{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      soc.s.Name,
-				Namespace: soc.s.Namespace,
-				Labels:    soc.s.Labels,
-				OwnerReferences: []metav1.OwnerReference{
-					*metav1.NewControllerRef(soc.s, v1alpha1.SchemaGroupVersionKind),
-				},
-			},
-			Spec: *soc.s.Spec.DeploySpec,
+	// expose sensor if service is configured
+	if soc.getServiceSpec() != nil {
+		svc, err := soc.createSensorService()
+		if err != nil {
+			soc.log.Error().Err(err).Str("sensor-name", soc.s.Name).Msg("failed to create service for sensor")
+			soc.markSensorPhase(v1alpha1.NodePhaseError, false, fmt.Sprintf("failed to create sensor service. err: %s", err))
+			return err
 		}
-		_, err = soc.controller.kubeClientset.CoreV1().Pods(soc.s.Namespace).Create(sensorPod)
+		soc.log.Info().Str("svc-name", svc.Name).Msg("sensor service is created")
+	}
+
+	// if we get here - we know the signals are running
+	soc.log.Info().Msg("marking sensor as active")
+	soc.markSensorPhase(v1alpha1.NodePhaseActive, false, "listening for events")
+	return nil
+}
+
+func (soc *sOperationCtx) updateSensorResources() error {
```

Comment:
Similar to gateway update method, I am proposing https://gist.github.com/VaibhavPage/0ba7d55df65487d590cf550075797936